### PR TITLE
Adds ADIOS2, high performance parallel I/O backend  to WRF [after fixes]

### PR DESCRIPTION
**Pull request by MichaelLaufer**
Optional high-performance parallel-I/O option added to WRF, enabling in-situ analysis using the ADIOS2 library.

This PR fixes changes requested in #1666 (commits were squashed to a single commit) , and is now rebased on the current development branch.  

*[Copied from #1666]*

<blockquote>
Optional high-performance parallel-I/O option added to WRF, enabling in-situ analysis using the ADIOS2 library.

**TYPE:** 
new feature

**KEYWORDS:** 
ADIOS2, In-situ, I/O, Parallel, High-performance

**SOURCE:** 
  Michael Laufer (Toga Networks, a Huawei Company)
  Erick Fredj, (The Jerusalem College of Technology, Rutgers University, Toga Networks, a Huawei Company) 
  Joseph Brodie (Rutgers University)
  Lori Garzio (Rutgers University)

**DESCRIPTION OF CHANGES:**
This PR adds [ADIOS2 ](https://github.com/ornladios/ADIOS2) as a high-performance parallel I/O backend option to WRF. ADIOS2 is a data management library that can transport groups of self-describing variables and attributes across different media types (such as files, wide-area-networks, and remote direct memory access). See [here ](https://doi.org/10.1016/j.softx.2020.100561) for additional information on ADIOS2.
Our testing demonstrates an order of magnitude performance increase over PnetCDF and can even outperform split NetCDF (io_form_* 102) at scale without the need to merge the files back together, while attaining compression ratios similar to NetCDF4 compression. ADIOS2 additionally supports in-situ processing and code-coupling without requiring data to traverse via the file system. A complete description of this new I/O backend, as well as some of the new features now accessible in WRF, can be found in our research article ([https://arxiv.org/abs/2201.08228](https://arxiv.org/abs/2201.08228)).
A [conversion script](https://github.com/MichaelLaufer/WRF-ADIOS2-to-NetCDF4) converting the ADIOS2 file format to the NetCDF4 file format is also provided for backward compatibility.
Details on configuring and running should be found in the file doc/README.adios2.

**LIST OF MODIFIED FILES:**
M Registry/Registry.CONVERT
M Registry/Registry.EM_COMMON
M Registry/Registry.EM_COMMON.var
M Registry/registry.io_boilerplate
M arch/Config.pl
M arch/postamble
M arch/preamble
M configure
M external/Makefile
A external/io_adios2/Makefile
A external/io_adios2/ext_adios2_get_dom_ti.code
A external/io_adios2/ext_adios2_get_var_td.code
A external/io_adios2/ext_adios2_get_var_ti.code
A external/io_adios2/ext_adios2_put_dom_ti.code
A external/io_adios2/ext_adios2_put_var_td.code
A external/io_adios2/ext_adios2_put_var_ti.code
A external/io_adios2/field_routines.F90
A external/io_adios2/transpose.code
A external/io_adios2/wrf_io.F90
M external/ioapi_share/wrf_status_codes.h
M frame/md_calls.m4
M frame/module_io.F
M share/module_io_domain.F
M share/output_wrf.F
M share/wrf_ext_write_field.F

**TESTS CONDUCTED:**
- Gnu and Intel compilers were used for testing with dmpar and dm+sm. The serial configuration is not applicable, as MPI is required for this implementation.
- Restarting WRF via the ADIOS2 restart file works properly. The data in subsequent output files matches the data in NetCDF output files.

**RELEASE NOTE:**
ADIOS2 has been added to WRF as an optional high-performance I/O backend. File write times are substantially lower than PnetCDF, with compression ratios close to NetCDF4. Allows for new in-situ processing and code-coupling capabilities through the ADIOS2 library.
</blockquote>

### CODE REVIEW

**Comment by weiwangncar** [comment]
The Jenkins tests have passed:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           58
    Number of Simulations  : 158           156        0
    Number of Comparisons  : 95           92        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

File: arch/Config.pl
Comment (mgduda) [inline arch/Config.pl:714]: When I install ADIOS2, the libraries are placed in a `lib64` directory rather than a `lib` directory. I recall on PR #1666 that @fengggli had found that their libraries were installed in a `lib` directory, though. Perhaps there needs to be some additional instruction on how to install ADIOS2, or -- better yet, in my opinion -- we could add some logic to determine where the ADIOS2 libraries are located?

**Review by mgduda — CHANGES_REQUESTED** [review]

**Comment by mgduda** [comment]
@MichaelLaufer Could you transfer (i.e., copy-and-paste) the PR description from #1666 to this PR? I believe we use the PR description as the commit message when merging a PR, so having the full description in this PR will be important.

File: arch/Config.pl
Comment (fengggli) [reply arch/Config.pl:714]: @mgduda I installed ADIOS2 2.8.0 on two machines, on my Ubuntu20.04 it's in /lib, but on another HPC system (PSC Bridges2, with CentOS 8.2) it's in /lib64.

@MichaelLaufer  ADIOS2 installation comes with a `adios2-config` tool at `ADIOS2-install-dir/bin/adios2-config`, and it shows the library path correctly for the above two environments. Below is an example on my Ubuntu (perhaps useful...):
```bash
fengggli@DESKTOP-B450FL(:):install/bin/adios2-config -f
-DADIOS2_USE_MPI -I/home/fengggli/Workspace/insituLauncher/install/include/adios2/fortran -Wl,-rpath,/home/fengggli/Workspace/insituLauncher/install/lib /home/fengggli/Workspace/insituLauncher/install/lib/libadios2_fortran_mpi.so.2.8.0 /home/fengggli/Workspace/insituLauncher/install/lib/libadios2_fortran.so.2.8.0 -Wl,-rpath-link,/home/fengggli/Workspace/insituLauncher/install/lib
fengggli@DESKTOP-B450FL(:):install/bin/adios2-config --fortran-flags
-DADIOS2_USE_MPI -I/home/fengggli/Workspace/insituLauncher/install/include/adios2/fortran
fengggli@DESKTOP-B450FL(:):install/bin/adios2-config --fortran-libs
-Wl,-rpath,/home/fengggli/Workspace/insituLauncher/install/lib /home/fengggli/Workspace/insituLauncher/install/lib/libadios2_fortran_mpi.so.2.8.0 /home/fengggli/Workspace/insituLauncher/install/lib/libadios2_fortran.so.2.8.0 -Wl,-rpath-link,/home/fengggli/Workspace/insituLauncher/install/lib
```

Comment (MichaelLaufer) [reply arch/Config.pl]: @mgduda and @fengggli thank you for the valuable feedback. 
I have just pushed a commit that addresses this by adding the `/lib64` path in addition to `/lib` for the ADIOS2 lib path. In my testing this works as expected.

When, and if this makes it to mainline WRF I will update the Spack WRF package that will streamline this process further. A simple `spack install wrf +adios2` is all that will be required. I have also tested this locally and it works smoothly.

**Comment by MichaelLaufer** [comment]
> @MichaelLaufer Could you transfer (i.e., copy-and-paste) the PR description from #1666 to this PR? I believe we use the PR description as the commit message when merging a PR, so having the full description in this PR will be important.

Done

**Review by mgduda — CHANGES_REQUESTED** [review]
Although I'm not able to trigger this behavior with the GNU 10.1.0 compilers, I seem to recall that there are (or were?) compilers that would complain if a non-existent library search path was given. It might therefore be preferable to try to detect whether the ADIOS2 installation uses `/lib` or `/lib64`. I've got a commit ([cfadf964](https://github.com/mgduda/WRF/commit/cfadf96451a42bda75f06a4dbd95a3a9a1523929)) in my fork of the WRF repository with some proposed changes for doing this. If these look fine to you (and assuming I have permissions to do so), I can push that commit onto the head of this PR branch. But, if you'd like to take a different approach, that would be fine, too.

Otherwise, I think this PR is in good shape.

**Comment by MichaelLaufer** [comment]
Thank you @mgduda, your solution seems optimal and appreciate your assistance. Please feel free to push that commit and proceed.

**Review by mgduda — APPROVED** [review]

**Comment by weiwangncar** [comment]
The Jenkins test results:

Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           57
    Number of Simulations  : 158           150        0
    Number of Comparisons  : 95           86        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

**Review by mgduda — APPROVED** [review]

**Comment by mgduda** [comment]
@weiwangncar I think this PR needs a second approval before it can be merged. Do you have any suggestions for reviewers?

**Comment by dudhia** [comment]
I looked at it and saw no red flags.

**Comment by weiwangncar** [comment]
@mgduda @dudhia I don't have any problem with Jimy approving this PR.

**Pull request merged**
